### DISCUSSION TITLE
Built with sdrplay option, RSPdx not detected.

### DISCUSSION THREAD

**Discussion by loughkb** [AlexandreRouma/SDRPlusPlus#356] [Q&A]
Howdy,

    I just rebuilt from source on linux, including the sdrplay support.   I used:
cmake .. -OPT_BUILD_SDRPLAY_SOURCE

It seemed to build fine, but my SDRPlay RSPdx does not appear in the sources list.   GQRX does see it fine and it works there.

Did I do it wrong?

**Comment by AlexandreRouma** [comment]
you need to do `-DOPT_BUILD_SDRPLAY_SOURCE=ON`, not `-OPT_BUILD_SDRPLAY_SOURCE` (notice the missing `=ON` and the `D` at the beginning)

**Comment by loughkb** [comment]
Thanks Alexandre,
   I rebuilt using the option and it seemed to build fine, including the sdrplay module.
   However, it still doesn't seem to detect it.   I looked at the module manager and don't see the sdrplay module listed.   I can find it in the drop down menu, but clicking the plus sign doesn't seem to add it.
   It is detected under soapy, but clicking the play button just hangs and the console outputs the following error:
   [ERROR] error in activateStream() - Init() failed: sdrplay_api_Fail

However,  in GQRX, it does work fine.

**Reply by AlexandreRouma** [reply]
GQRX uses it's own instance of the API, it's not comparable.

Make sure you've closed all other SDR apps (including GQRX) and restart the SDRplay API service.

**Reply by AlexandreRouma** [reply]
Also, please send your startup logs. If you updated from a very old version without the sdrplay module it might not be getting loaded.

PS: is there a reason you're building it yourself rather than using a prebuilt version?

**Comment by loughkb** [comment]
Well, I'm building from source to make sure everything is up to date.  Also to gather data for my next video.  I'm going to walk people through building it on the Raspberry Pi and then publish a script to automate building it on the Pi.  (Since there's no .deb provided for the Pi, I figure that will be useful for Pi users.)

Here's the startup.
loughkb@darter-pro:~$ sdrpp
[2021-09-13 07:47:39.184] [info] SDR++ v1.0.3
[2021-09-13 07:47:39.185] [info] Loading config
[2021-09-13 07:47:39.333] [info] Loading icons
[2021-09-13 07:47:39.338] [info] Loading band plans
[2021-09-13 07:47:39.339] [info] Loading band plans color table
[2021-09-13 07:47:39.342] [info] Loading modules
[2021-09-13 07:47:39.342] [info] Loading /usr/lib/sdrpp/plugins/meteor_demodulator.so
[2021-09-13 07:47:39.343] [info] Loading /usr/lib/sdrpp/plugins/discord_integration.so
[2021-09-13 07:47:39.343] [info] Loading /usr/lib/sdrpp/plugins/frequency_manager.so
[2021-09-13 07:47:39.358] [info] Loading /usr/lib/sdrpp/plugins/airspy_source.so
[2021-09-13 07:47:39.375] [info] Loading /usr/lib/sdrpp/plugins/airspyhf_source.so
[2021-09-13 07:47:39.392] [info] Loading /usr/lib/sdrpp/plugins/rtl_tcp_source.so
[2021-09-13 07:47:39.408] [info] Loading /usr/lib/sdrpp/plugins/soapy_source.so
[2021-09-13 07:47:39.425] [info] Loading /usr/lib/sdrpp/plugins/file_source.so
[2021-09-13 07:47:39.442] [info] Loading /usr/lib/sdrpp/plugins/network_sink.so
[2021-09-13 07:47:39.458] [info] Loading /usr/lib/sdrpp/plugins/hackrf_source.so
[2021-09-13 07:47:39.475] [info] Loading /usr/lib/sdrpp/plugins/rtl_sdr_source.so
[2021-09-13 07:47:39.492] [info] Loading /usr/lib/sdrpp/plugins/radio.so
[2021-09-13 07:47:39.509] [info] Loading /usr/lib/sdrpp/plugins/plutosdr_source.so
[2021-09-13 07:47:39.527] [info] Loading /usr/lib/sdrpp/plugins/weather_sat_decoder.so
[2021-09-13 07:47:39.542] [info] Loading /usr/lib/sdrpp/plugins/spyserver_source.so
[2021-09-13 07:47:39.558] [info] Loading /usr/lib/sdrpp/plugins/rigctl_server.so
[2021-09-13 07:47:39.575] [info] Loading /usr/lib/sdrpp/plugins/recorder.so
[2021-09-13 07:47:39.592] [info] Loading /usr/lib/sdrpp/plugins/sdrplay_source.so
[2021-09-13 07:47:39.608] [info] Loading /usr/lib/sdrpp/plugins/audio_sink.so
[2021-09-13 07:47:39.628] [info] Initializing Airspy Source (airspy_source)
[2021-09-13 07:47:39.652] [info] Initializing AirspyHF+ Source (airspyhf_source)
[2021-09-13 07:47:39.662] [info] Initializing Audio (audio_sink)
[2021-09-13 07:47:39.674] [info] Initializing File Source (file_source)
[2021-09-13 07:47:39.691] [info] Initializing Frequency Manager (frequency_manager)
[2021-09-13 07:47:39.707] [info] Initializing HackRF Source (hackrf_source)
[2021-09-13 07:47:39.729] [info] Initializing Meteor (meteor_demodulator)
[2021-09-13 07:47:39.758] [info] Initializing RTL-SDR Source (rtl_sdr_source)
[2021-09-13 07:47:39.760] [info] Initializing RTL-TCP Source (rtl_tcp_source)
[2021-09-13 07:47:39.774] [info] Initializing Radio (radio)
[2021-09-13 07:47:39.834] [info] RtAudio stream open
[2021-09-13 07:47:39.834] [info] Initializing Recorder (recorder)
[2021-09-13 07:47:39.834] [info] Initializing Rigctl Server (rigctl_server)
[2021-09-13 07:47:39.837] [info] Initializing SoapySDR Source (soapy_source)
[INFO] [UHD] linux; GNU C++ version 9.2.1 20200304; Boost_107100; UHD_3.15.0.0-2build5
libusb: error [udev_hotplug_event] ignoring udev action bind
[INFO] devIdx: 0
[INFO] hwVer: 4
[INFO] rspDuoMode: 0
[INFO] tuner: 1
[INFO] rspDuoSampleFreq: 0.000000
[ERROR] SelectDevice Error: sdrplay_api_Fail
terminate called after throwing an instance of 'std::runtime_error'
  what():  SelectDevice() failed
Aborted (core dumped)

**Reply by AlexandreRouma** [reply]
Looks like the module isn't initialized, you likely have an old config. I'd suggest completely resetting the configuration by deleting ~/.config/sdrpp

> then publish a script to automate building it on the Pi. (Since there's no .deb provided for the Pi, I figure that will be useful for Pi users.)

There already a script in the root of the project to do that.

**Comment by loughkb** [comment]
By the way, GQRX wasn't running and this was after a reboot.

**Comment by loughkb** [comment]
I did remove the config and it still didn't work with an API fail message.
Just for the heck of it, I re-installed the API and now it works.   Weird!   I don't know what could have changed since I originally installed it, it's the same version number.   
Oh well, it's working, that's what matters.   Thanks for your help.

**Comment by loughkb** [comment]
Oh, I went and looked at the pi script, cool!   I'll probably still do a video though.  I think it will be helpful for folks to have a walkthrough.

**Reply by AlexandreRouma** [reply]
Would love some feedback on the script. I haven't yet been able to try it out as my raspberry is broken so I kinda wrote it blind.

**Comment by loughkb** [comment]
When I finish the video, I'll send you a link.   I might post it here in the General section if that's ok.

**Reply by AlexandreRouma** [reply]
I assume from the username that you're Kevin Loughin right?
Saw your video about it :)

I'll was gonna comment some stuff about the issues you encountered, thanks a lot for the feedback!

**Comment by loughkb** [comment]
Yep, it's me.   I wondered if you'd spot the vid when I published it.
I'll fill in a bit about the solution to this issue at the head of the rasp pi vid.

**Reply by AlexandreRouma** [reply]
I think what was happening is you had a config from one of the old versions without support for the sdrplay and it wasn't intiialising the sdprlay_source mùodule. You can see in the logs, it loads the module (so it's installed), but it's not getting to the "Initializing..." part.
The reboot probably also helped. The SDRplay API is horribly broken on linux, I've been in contact with SDRplay about it but I doubt we'll see a solutions soon. Usuually what fixes issues for me is restarting the sdrplay service.

Above the "bullseye" mode. This is center tuning. It's currently kinda broken and I'll try to fix it soon.

I also noticed you were missing a few other modules before resetting the config (for example the network sink). This adds similar functionality to GQRX to stream audio or baseband via UDP or TCP (both in server and client mode). This is pretty useful for decoding digital modes (I posted an example on twitter of decoding multiple pagers at once using it).

Speaking of which, I'm not sure if you noticed, but you can add as many demodulators as you want at once. You simply need to add additional instances of the radio module through the Module Manager menu. Same goes for the recorder, frequency manager and rigctl server.

**Reply by AlexandreRouma** [reply]
FYI I pushed a fix for the center tuning mode.

**Comment by loughkb** [comment]
Hello again.
   Working on the video and script today.  I ran into a couple of things.
   First, libvolk2-dev wasn't available in the repository.  My raspian image is over a year old, so it may be a version back from current, not sure.   libvolk1-dev was available and cmake didn't complain, so I'm attempting the build to see what happens.  It's compiling as I write this.
   Second, I added a step to clone your git and change into that directory before the step that makes the build folder.
   Additionally, I added the option to build in the SDRplay support.   From comments and feedback on the previous video, I gather that quite a few folks have the SDRplay hardware, so maybe that should be a default instead of an option?   Maybe add a bumper text at the beginning of the script to remind users that if they have the SDRplay, they need to install the API first.

If libvolk1 is incompatible and it doesn't work,  I'll have to download the latest raspian and start over.  I'll let you know.    If it does work, maybe add a check on the raspian version that downloads the appropriate version for the users OS version.

**Reply by AlexandreRouma** [reply]
> libvolk1-dev was available and cmake didn't complain

Yes, that works too, but note that I might have to drop support for volk 1.3 and lower later in development. Volk 1.4 should however always be supported. Hopefully that's what even old RPi images use. Debian based distros tend to be extremely out of date and that has caused me quite a few issues with the project...

> Second, I added a step to clone your git and change into that directory before the step that makes the build folder.

Soon, installing should be as easy as running the script provided with the source code (I'll add an option to select drivers)
In the mean time, feel free to send over the instructions you're planning to give so I can make sure they're correct.

> so maybe that should be a default instead of an option?

The reason the SDRplay isn't enabled by default is it's library is proprietary and cannot be installed through most package manager

**Reply by pmaine** [reply]
I used libvolk1-dev on a new pi build and the build worked perfectly.

**Comment by loughkb** [comment]
Got it.

   The first build attempt failed on the bladeRF stuff.   The cascade of errors started with this:
   Scanning dependencies of target bladerf_source
[ 56%] Building CXX object bladerf_source/CMakeFiles/bladerf_source.dir/src/main.cpp.o
/home/loughkb/Desktop/SDRPlusPlus/bladerf_source/src/main.cpp:523:11: error: \u2018bladerf_range\u2019 does not name a type; did you mean \u2018bladerf_image\u2019?
     const bladerf_range* srRange = NULL;

I removed that option from the cmake line and restarted the build.  Compiling now, it takes awhile.

**Reply by AlexandreRouma** [reply]
Looks like the pi has an outdated version of libbladerf

**Comment by loughkb** [comment]
It's working.   A little bit of crackle to the audio occasionally, but it worked with both my SDRPlay and Airspy HF + discovery.  

Below is the script that works, with my modifications.   In the video I cover each change and why, I also mention that some of the issues may go away with newer versions of raspian.  Finally, I direct people to check your script on the github page and just use it once you've updated it.
(I don't know why git want's to change fonts and stuff when reading it. oh well.  Wait, I see..  Every # comment got bold and larger fonts.)
---------------- modified script below ---------------
#!/bin/sh
# Script authored by Alexandre Rouma, tested and modified by Kevin Loughin, KB9RLW, Sept 15th 2021
set -e

[ $(id -u) = 0 ] && echo "Please do not run this script as root" && exit 100

echo "Installing dependencies"
sudo apt update
sudo apt install -y build-essential cmake git libfftw3-dev libglfw3-dev libglew-dev libvolk1-dev libsoapysdr-dev libairspyhf-dev libairspy-dev \
            libiio-dev libad9361-dev librtaudio-dev libhackrf-dev librtlsdr-dev libbladerf-dev liblimesuite-dev p7zip-full wget
# get the source
cd ~/Desktop
git clone https://github.com/AlexandreRouma/SDRPlusPlus
cd SDRPlusPlus

echo "Preparing build"
mkdir -p build
cd build
cmake .. -DOPT_BUILD_LIMESDR_SOURCE=ON -DOPT_BUILD_SDRPLAY_SOURCE=ON

echo "Building"
make

echo "Installing"
sudo make install

echo "Done!"

**Reply by AlexandreRouma** [reply]
Looks good, just make sure to mention that on newer raspbian versions users might need to use libvolk2-dev and it should be perfect.

**Comment by loughkb** [comment]
Ah, recorded audio from the program was clean without crackling.  I think that's an external issue with my HDMI audio to the monitor.  Disregard that comment about the audio.

**Comment by loughkb** [comment]
I did mention that, and a few other things.
The video is in preview until Monday,  but if you want to check it out, here's the link.
https://youtu.be/Cjf9P8a1uTQ

**Reply by AlexandreRouma** [reply]
Thanks! I'll have a look

**Reply by AlexandreRouma** [reply]
Alright so, I noticed you added the git clone in the script.

How you're actually supposed to do it is, you clone the repo like usual, and then you just run the script from inside the repo ;)

So it would be;

```sh
git clone https://github.com/AlexandreRouma/SDRPlusPlus
cd SDRPlusPlus
chmod +x ./rpi_install.sh
./rpi_install.sh
```

EDIT: Should've noticed that when you sent the script, sorry

I'll try to add detection for older raspbian versions so it uses libvolk1 instead when needed

**Comment by loughkb** [comment]
Well, my goal is to create a simple single script so that the user only has to download one simple item and run it.  Make it as easy as possible for the end user.  The less they have to do, the happier they are and the better your project will do.
I worked in computer support for over 30 years, so I tend to think about the users point of view when I'm doing things.
I'd create a .deb for the Pi if I knew how so it could be included on the releases page.  That would be the best solution in my opinion.
Maybe you know someone with a pi that could help with that?

**Reply by AlexandreRouma** [reply]
To create a deb, just run `sh make_debian_package.sh ./build 'libfftw3-dev, libglfw3-dev, libglew-dev, libvolk1-dev, librtaudio-dev'` inside the SDRPlusPlus directory. Should drop the created deb in the same directory
